keps/sig-node/3327-align-by-socket/README.md

```diff
@@ -0,0 +1,252 @@
+
+### Non-Goals
+  * Guarantee optimal NUMA allocation for cpu distribution.
+
+## Proposal
+
+We propose to add a new CPUManager policy option called align-by-socket to the static CPUManager policy. With this policy, the CPUManager will prefer those hints which are within the same socket (as opposed to just within the same NUMA node) if it is possible to have all CPUs allocated from the same socket.
+
+### Risks and Mitigations
+
+The risks of adding this new feature are quite low.
+It is isolated to a specific policy option within the `CPUManager`, and is protected both by the option itself, as well as the `CPUManagerPolicyOptions` feature gate (which is disabled by default).
+
+| Risk                                             | Impact | Mitigation |
+| -------------------------------------------------| -------| ---------- |
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but CPU allocations can spread across socket in cases when allocation could have been within same socket |
+
+## Design Details
+
+### Proposed Change 
+
+When align-by-socket is enabled as a policy option,  the CPUManager’s GetTopologyHints() function will generate hints based on the sockets that a group of CPUs  belong to, rather than the NUMA nodes they belong to.
+
+To achieve this, the following updates are needed to the GetTopologyHints() function:
+```
+func (p *staticPolicy) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, reusableCPUs cpuset.CPUSet, request int) []topologymanager.TopologyHint {
+	...
+
+	// Loop back through all hints and update the 'Preferred' field based on
+	// counting the number of bits sets in the affinity mask and comparing it
+	// to the minAffinitySize. Only those with an equal number of bits set (and
+	// with a minimal set of numa nodes) will be considered preferred.
+	for i := range hints {
+		if p.options.AlignBySocket && isSocketAligned(hints[i].NUMANodeAffinity) {
+			hints[i].Preferred = true
+			continue
+		}
+		if hints[i].NUMANodeAffinity.Count() == minAffinitySize {
+			hints[i].Preferred = true
+		}
+	}
+
+	return hints
+}
+
+```
+At the end, we will have a list of desired hints. These hints will then be passed to the topology manager whose job it is to select the best hint (with an increased likelihood of selecting a hint that has CPUs which are aligned by socket now).
+
+In case TopologyManager “single-numa-node” policy is enabled, the policy option of “align-by-socket” is redundant since allocation guarantees within the same numa are by definition socket aligned. Hence, we will error out in case the policy option of “align-by-socket” is enabled in conjunction with TopologyManager single-numa-node policy.
```

Comment:
This is something, i will need to think more!
Let me think on situation where a single NUMA can span multiple sockets and update once i have some definitive answer

Reply:
Can we please have few examples of architectures we are targeting where a single NUMA can span multiple sockets?
[I'm aware such architecture existed in the past, but I'm not sure the hardware is heading in this direction](https://github.com/kubernetes/kubernetes/blob/v1.24.1/pkg/kubelet/cm/cpumanager/topology/topology_test.go#L238).

Reply:
Updated the KEP, i am leaning towards erring out in case there are more than one socket per NUMA
We can address this use case if we see real world architecture leaning in that direction.
I am open to suggestions

Reply:
I agree with @klueska and I prefer approach number 2 as well. Would be fine by me.